### ISSUE TITLE
Repository submodule changes are not getting detected/fetched

### ISSUE THREAD

**Issue by hristorikev** [concourse/git-resource#375]
Hi all,

We are using an older version of Concourse (v6.5.1 to be specific), hence the build-in resource versions are older. While working on upgrading our instance we encountered a problem where the git resource was not detecting changes in it's submodules.

I did some more testing and I found that this behaviour came with v1.12.0 of the resource and it's still relevant with the latest version as well. In v1.11.0 this was not the case, the git resource did not have the same behaviour.

To explain better I made a public setup with a pipeline, which anyone can take and test it.

This is my "parent" repository: https://github.com/hristorikev/parent_repo
In that repository I have added under the "children" folder a submodule of the "child" repository: https://github.com/hristorikev/child_repo.

This is our example pipeline setup:

```
---
resources:
- name: submodule-test-repo
  type: git
  icon: github
  check_every: 2m
  source:
    uri: https://github.com/hristorikev/parent_repo
    branch: main
    paths:
      - config/*
      - config/**/*
      - children/child_repo/*
      - children/child_repo/**/*

resource_types:
- name: git
  type: registry-image
  source:
    repository: concourse/git-resource
    tag: 1.12.0

jobs:
- name: job
  public: true
  plan:
  - get: submodule-test-repo
    trigger: true
  - task: list-files
    config:
      inputs:
        - name: submodule-test-repo
      platform: linux
      image_resource:
        type: registry-image
        source:
          repository: taylorsilva/carvel-ytt
      run:
        path: ls
        args: ["-la", "./submodule-test-repo/children/child_repo"]
```

If you check the 5 commits in the parent repository:

  1. e1029600d6ee1c432915aef8184ef066c2b66e96
  2. 82735f6fc57fa4a67206193609f85d2efb55a217
  3. 5a62a8e39e32693d310b6bca77b87a649b8e3e02
  4. 2b6f5af8158c71ba43e79c0c894196794fb032f7
  5. 9ee033e8dced0f2f50dd4099281d79991fb26fb8
  
The first two commits are changes in the parent repository, and commit 3, 4 and 5 are changes in the submodule repository.  With v1.11.0 all changes were detected and the resource check detected the latest commit ```9ee033e8dced0f2f50dd4099281d79991fb26fb8```, but with the example pipeline above with the newer version of the git-resource only the changes in the parent repo are detected ```2b6f5af8158c71ba43e79c0c894196794fb032f7``` . 
 _This is also the behaviour with v1.14.5_ 
 
Could someone help us understand what has changed? Is this a known issue? 
 
Please let me know if I need to give more details. Thank you!
 
Kind Regards,
Hristo

**Comment by hristorikev** [comment]
Opened a parallel discussion in the Concourse community regarding a way to possibly pin the resource version as a Concourse configuration, as this is still stopping us from upgrading our instance. We have too many pipelines which need to be manually re-configured.

https://github.com/concourse/concourse/discussions/7976

**Comment by hristorikev** [comment]
I looked deeper into the issue and it came out that it's not caused by the git-resource but from git itself. [A change was introduced in git v2.29.0 ](https://github.com/git/git/commit/35a9f1e99c5d31635bb78a4f2d498e72c04fc471). This change fixed an issue in git, which should have been fixed a long time ago, where it was treating submodules as folders and the wildcard feature worked the same. 

This is what happened:

The folllowing command:
`git rev-list --all --first-parent -- children/child_repo/* children/child_repo/**/* config/* config/**/*` 
is executed based on the configuration in the description of this issue and this in the git version above 2.29.0 returned only the following commits:

`2b6f5af8158c71ba43e79c0c894196794fb032f7`
`9ee033e8dced0f2f50dd4099281d79991fb26fb8`

But if we change the paths in the git rev-list command:
`git rev-list --all --first-parent -- children/child_repo config/* config/**/*` 

Returned all commits related to the submodule as well as expected:
`e1029600d6ee1c432915aef8184ef066c2b66e96`
`82735f6fc57fa4a67206193609f85d2efb55a217`
`5a62a8e39e32693d310b6bca77b87a649b8e3e02`
`2b6f5af8158c71ba43e79c0c894196794fb032f7`
`9ee033e8dced0f2f50dd4099281d79991fb26fb8`

So we updated our paths and now it should work fine with the newer version of the git-resource.
